homeassistant/components/stream/hls.py

```diff
@@ -0,0 +1,128 @@
+"""
+Provide functionality to stream HLS.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/stream/hls
+"""
+import datetime
+
+from aiohttp import web
+
+from .core import StreamView, StreamOutput
+
+
+async def async_setup_hls(hass):
+    """Set up api endpoints."""
+    hass.http.register_view(HlsPlaylistView())
+    hass.http.register_view(HlsSegmentView())
+    return '{}/api/hls/{}/playlist.m3u8'
+
+
+class HlsPlaylistView(StreamView):
+    """Stream view to serve a M3U8 stream."""
+
+    url = r'/api/hls/{token:[a-f0-9]+}/playlist.m3u8'
+    name = 'api:stream:hls:playlist'
+    cors_allowed = True
+
+    async def handle(self, request, stream, sequence):
+        """Return m3u8 playlist."""
+        renderer = M3U8Renderer(stream)
+        track = stream.add_provider(HlsStreamOutput())
+        stream.start()
+        # Wait for a segment to be ready
+        if not track.segments:
+            await track.recv()
+        headers = {
+            'Content-Type': 'application/vnd.apple.mpegurl'
+        }
+        return web.Response(body=renderer.render(
+            track, datetime.datetime.utcnow()
+            ).encode("utf-8"), headers=headers)
+
+
+class HlsSegmentView(StreamView):
+    """Stream view to serve a MPEG2TS segment."""
+
+    url = r'/api/hls/{token:[a-f0-9]+}/segment/{sequence:\d+}.ts'
+    name = 'api:stream:hls:segment'
+    cors_allowed = True
+
+    async def handle(self, request, stream, sequence):
+        """Return mpegts segment."""
+        track = stream.add_provider(HlsStreamOutput())
+        segment = track.get_segment(int(sequence))
+        if not segment:
+            return web.HTTPNotFound()
+        headers = {
+            'Content-Type': 'video/mp2t'
+        }
+        return web.Response(body=segment.segment.getvalue(), headers=headers)
+
+
+class M3U8Renderer:
+    """M3U8 Render Helper."""
+
+    def __init__(self, stream, version=3):
+        """Initialize renderer."""
+        self.stream = stream
+        self.version = version
+
+    def render_preamble(self, track):
+        """Render preamble."""
+        return [
+            "#EXT-X-VERSION:{}".format(self.version),
+            "#EXT-X-TARGETDURATION:{}".format(track.target_duration),
+        ]
+
+    def render_segment(self, segment):
+        """Render segment."""
+        return [
+            "#EXTINF:{:.04},".format(float(segment.duration)),
+            "/api/hls/{}/segment/{}.ts".format(
```

Comment:
Could this be changed to be `./segment/{}.ts`, making it relative to the playlist url?

Reply:
Looks like it can be.